File: mqtt/solar/schedule.py
"""夜間暫停（Feature #4）。

零依賴的日出日落計算（NOAA 簡化版）。準確度 ±2 分鐘，對判斷白天 / 夜晚綽綽有餘。
"""

from __future__ import annotations

import math
from datetime import date, datetime, time, timedelta, timezone


def _solar_event(
    d: date,
    lat: float,
    lon: float,
    rising: bool,
    tz_offset_hours: float = 8.0,
) -> time | None:
    """回傳當地時間的日出 (rising=True) 或日落 (rising=False)。

    NOAA 簡化演算法。無日出 / 無日落的極地情況回 None。
    """
    # 儒略日
    n = d.toordinal() - date(2000, 1, 1).toordinal() + 1
    j_star = n - lon / 360.0 + (6.0 if rising else 0.0)
    j_star_rising = n - lon / 360.0  # 用於太陽黃經

    # 平近點角
    m = (357.5291 + 0.98560028 * j_star) % 360.0
    m_rad = math.radians(m)

    # 中心方程
    c = 1.9148 * math.sin(m_rad) + 0.0200 * math.sin(2 * m_rad) + 0.0003 * math.sin(3 * m_rad)

    # 黃道經度
    lam = (m + c + 180.0 + 102.9372) % 360.0
    lam_rad = math.radians(lam)

    # 太陽正中天（transit）
    j_transit = 2451545.0 + j_star_rising + 0.0053 * math.sin(m_rad) - 0.0069 * math.sin(2 * lam_rad)

    # 赤緯
    sin_delta = math.sin(lam_rad) * math.sin(math.radians(23.44))
    delta = math.asin(sin_delta)

    # 時角
    lat_rad = math.radians(lat)
    cos_h = (
        math.sin(math.radians(-0.83)) - math.sin(lat_rad) * math.sin(delta)
    ) / (math.cos(lat_rad) * math.cos(delta))
    if cos_h > 1 or cos_h < -1:
        return None
    h = math.degrees(math.acos(cos_h))

    # 日出 / 日落的儒略日
    if rising:
        j_event = j_transit - h / 360.0
    else:
        j_event = j_transit + h / 360.0

    # 轉回 UTC datetime
    unix_days = j_event - 2440587.5
    dt_utc = datetime(1970, 1, 1, tzinfo=timezone.utc) + timedelta(days=unix_days)
    dt_local = dt_utc + timedelta(hours=tz_offset_hours)
    return dt_local.time().replace(microsecond=0)


def sunrise(d: date, lat: float, lon: float, tz: float = 8.0) -> time | None:
    return _solar_event(d, lat, lon, rising=True, tz_offset_hours=tz)


def sunset(d: date, lat: float, lon: float, tz: float = 8.0) -> time | None:
    return _solar_event(d, lat, lon, rising=False, tz_offset_hours=tz)


def is_daytime(
    now: datetime,
    lat: float,
    lon: float,
    padding_min: int = 30,
    tz: float = 8.0,
) -> bool:
    """是否在「日出-padding ~ 日落+padding」之內。"""
    sr = sunrise(now.date(), lat, lon, tz)
    ss = sunset(now.date(), lat, lon, tz)
    if sr is None or ss is None:
        return True  # 無法判斷 → 保守視為白天
    sr_dt = datetime.combine(now.date(), sr) - timedelta(minutes=padding_min)
    ss_dt = datetime.combine(now.date(), ss) + timedelta(minutes=padding_min)
    return sr_dt <= now.replace(tzinfo=None) <= ss_dt


def seconds_until_sunrise(
    now: datetime,
    lat: float,
    lon: float,
    padding_min: int = 30,
    tz: float = 8.0,
) -> int:
    """回傳距離下個日出-padding 還有幾秒（夜間暫停時用）。最少 60。"""
    # 今天 or 明天的日出
    today = now.date()
    for offset in (0, 1):
        d = today + timedelta(days=offset)
        sr = sunrise(d, lat, lon, tz)
        if sr is None:
            continue
        target = datetime.combine(d, sr) - timedelta(minutes=padding_min)
        delta = (target - now.replace(tzinfo=None)).total_seconds()
        if delta > 0:
            return max(60, int(delta))
    return 3600  # 極區備援

File: solar_mqtt_go/internal/schedule/schedule.go
// Package schedule 提供夜間暫停用日出日落計算（NOAA 簡化版，零依賴）。
// 演算法逐行對照 Python solar/schedule.py；準確度 ±2 分鐘。
// 時間語意：輸入輸出皆為 naive 當地時間；tz 為當地時區相對 UTC 的偏移小時數（預設 8.0）。
package schedule

import (
	"math"
	"time"
)

const secondsPerDay = 86400

// unixDaysFromCivil 儒略日換算用：從 1970-01-01 起算的天數（UTC 正午對齊）。
func unixDaysFromCivil(y int, m time.Month, d int) int {
	return int(time.Date(y, m, d, 0, 0, 0, 0, time.UTC).Unix() / secondsPerDay)
}

// solarEvent 對應 Python _solar_event。回傳當地時間的日出/日落時刻（當天日期 + 時分秒），
// 極地無事件時回 nil。
func solarEvent(date time.Time, lat, lon float64, rising bool, tzOffsetHours float64) *time.Time {
	// 儒略日：n = toordinal(d) - toordinal(2000-01-01) + 1
	n := unixDaysFromCivil(date.Year(), date.Month(), date.Day()) -
		unixDaysFromCivil(2000, 1, 1) + 1

	lonTerm := float64(n) - lon/360.0
	jStar := lonTerm
	if rising {
		jStar = lonTerm + 6.0
	}

	// 平近點角
	mAng := math.Mod(357.5291+0.98560028*jStar, 360.0)
	mRad := deg2rad(mAng)

	// 中心方程
	c := 1.9148*math.Sin(mRad) + 0.0200*math.Sin(2*mRad) + 0.0003*math.Sin(3*mRad)

	// 黃道經度
	lam := math.Mod(mAng+c+180.0+102.9372, 360.0)
	lamRad := deg2rad(lam)

	// 太陽正中天
	jTransit := 2451545.0 + lonTerm + 0.0053*math.Sin(mRad) - 0.0069*math.Sin(2*lamRad)

	// 赤緯
	sinDelta := math.Sin(lamRad) * math.Sin(deg2rad(23.44))
	delta := math.Asin(sinDelta)

	// 時角
	latRad := deg2rad(lat)
	cosH := (math.Sin(deg2rad(-0.83)) - math.Sin(latRad)*math.Sin(delta)) /
		(math.Cos(latRad) * math.Cos(delta))
	if cosH > 1 || cosH < -1 {
		return nil
	}
	h := rad2deg(math.Acos(cosH))

	var jEvent float64
	if rising {
		jEvent = jTransit - h/360.0
	} else {
		jEvent = jTransit + h/360.0
	}

	// 轉回 UTC datetime 再加時區偏移
	unixDays := jEvent - 2440587.5
	dtUTC := time.Unix(int64(unixDays*secondsPerDay), 0).UTC()
	dtLocal := dtUTC.Add(time.Duration(tzOffsetHours * float64(time.Hour)))

	// 取「時分秒」（Python .time().replace(microsecond=0)），日期掛回輸入日
	tod := time.Date(date.Year(), date.Month(), date.Day(),
		dtLocal.Hour(), dtLocal.Minute(), dtLocal.Second(), 0, date.Location())
	return &tod
}

func deg2rad(d float64) float64 { return d * math.Pi / 180.0 }
func rad2deg(r float64) float64 { return r * 180.0 / math.Pi }

// Sunrise 回傳日期 d 的日出（當地時間，tz 預設 8.0）。
func Sunrise(d time.Time, lat, lon float64) *time.Time {
	return solarEvent(d, lat, lon, true, 8.0)
}

// Sunset 回傳日期 d 的日落。
func Sunset(d time.Time, lat, lon float64) *time.Time {
	return solarEvent(d, lat, lon, false, 8.0)
}

// IsDaytime 是否在「日出-padding ~ 日落+padding」之內。
// 無法判斷（極區）→ 保守視為白天。
func IsDaytime(now time.Time, lat, lon float64, paddingMin int) bool {
	sr := Sunrise(now, lat, lon)
	ss := Sunset(now, lat, lon)
	if sr == nil || ss == nil {
		return true
	}
	srDt := sr.Add(-time.Duration(paddingMin) * time.Minute)
	ssDt := ss.Add(time.Duration(paddingMin) * time.Minute)
	nowNaive := time.Date(now.Year(), now.Month(), now.Day(),
		now.Hour(), now.Minute(), now.Second(), 0, now.Location())
	return !nowNaive.Before(srDt) && !nowNaive.After(ssDt)
}

// SecondsUntilSunrise 回傳距離下個日出-padding 的秒數（最少 60；極區備援 3600）。
func SecondsUntilSunrise(now time.Time, lat, lon float64, paddingMin int) int {
	today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, now.Location())
	for offset := 0; offset <= 1; offset++ {
		d := today.AddDate(0, 0, offset)
		sr := Sunrise(d, lat, lon)
		if sr == nil {
			continue
		}
		target := sr.Add(-time.Duration(paddingMin) * time.Minute)
		delta := target.Sub(now).Seconds()
		if delta > 0 {
			s := int(delta)
			if s < 60 {
				s = 60
			}
			return s
		}
	}
	return 3600
}
